File: 2024/day15/day15.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strings"
)

//directions
var dirs = map[byte][2]int {
	'^': {-1, 0},
	'v': {1, 0},
	'<': {0, -1},
	'>': {0, 1},
}

func readGrid(filename string) [][]byte {
	file, err := os.Open(filename)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var grid [][]byte

	for scanner.Scan() {
		line := scanner.Text()
		grid = append(grid, []byte(line))
	}
	return grid
}

func readDirections(filename string) string{
	file, err := os.Open(filename)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	var moves strings.Builder

	for scanner.Scan() {
		moves.WriteString(scanner.Text())
	}

	return moves.String()
}

func move(grid [][]byte, r, c int, dir byte)(int , int) {
	dr := dirs[dir][0]
	dc := dirs[dir][1]

	nr := r + dr
	nc := c + dc

	// wall
	if grid[nr][nc] == '#' {
		return r, c
	}

	// empty
	if grid[nr][nc] == '.' {
		grid[r][c] = '.'
		grid[nr][nc] = '@'
		return nr, nc
	}

	// pushing box
	if grid[nr][nc] == 'O' {
		cr, cc := nr, nc
		
		// find end of the box chain
		for grid[cr][cc] == 'O' {	
			cr += dr
			cc += dc
		}	
		
		//blocked by wall
		if grid[cr][cc] == '#' {
			return r, c
		}

		// must be empty now
		if grid[cr][cc] == '.' {
			// move last box
			grid[cr][cc] = 'O'

			// move robot
			grid[nr][nc] = '@'
			grid[r][c] = '.'

			return nr, nc
		}
	}
	return r, c
}

func gpsSum(grid [][]byte) int {
	sum := 0

	for r := range grid {
		for c := range grid[r] {
			if grid[r][c] == 'O' {
				sum += 100*r + c
			}
		}
	}

	return sum
}

func main() {
	grid := readGrid("grid.txt")
	moves := readDirections("directions.txt")


	// first find the position of the robot
	var r, c int
	for i := range grid {
		for j := range grid[i] {
			if grid[i][j] == '@' {
				r, c = i, j
			}
		}
	}

	for i := 0; i < len(moves); i++ {
		r, c = move(grid, r, c, moves[i])
	}

	fmt.Println(gpsSum(grid))
}

File: 2024/day15/Day15.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class Day15 {

    static char[][] grid;
    static String moves;
    static int robotRow;
    static int robotCol;

    static Map<Character, int[]> directions = Map.of(
        '^', new int[]{-1, 0},
        'v', new int[]{1, 0},
        '<', new int[]{0, -1},
        '>', new int[]{0, 1}
    );

    static void readGrid() throws Exception {
        List<String> lines = new ArrayList<>();

        BufferedReader br = new BufferedReader(new FileReader("grid.txt"));

        String line;
        while ((line = br.readLine()) != null) {
            if (!line.isEmpty()) {
                lines.add(line);
            }
        }

        br.close();

        grid = new char[lines.size()][];

        for (int i = 0; i < lines.size(); i++) {
            grid[i] = lines.get(i).toCharArray();
        }
    }

    static void readMoves() throws Exception {
        BufferedReader br = new BufferedReader(new FileReader("directions.txt"));

        StringBuilder sb = new StringBuilder();

        String line;
        while ((line = br.readLine()) != null) {
            sb.append(line);
        }

        br.close();

        moves = sb.toString().replaceAll("[^<>^v]", "");
    }

    static void findRobot() {
        for (int r = 0; r < grid.length; r++) {
            for (int c = 0; c < grid[r].length; c++) {
                if (grid[r][c] == '@') {
                    robotRow = r;
                    robotCol = c;
                    return;
                }
            }
        }
    }

    static void move(char dir) {
        int[] d = directions.get(dir);

        int dr = d[0];
        int dc = d[1];

        int nr = robotRow + dr;
        int nc = robotCol + dc;

        if (grid[nr][nc] == '#') {
            return;
        }

        if (grid[nr][nc] == '.') {
            grid[robotRow][robotCol] = '.';
            grid[nr][nc] = '@';

            robotRow = nr;
            robotCol = nc;
            return;
        }

        if (grid[nr][nc] == 'O') {
            int br = nr;
            int bc = nc;

            while (grid[br][bc] == 'O') {
                br += dr;
                bc += dc;
            }

            if (grid[br][bc] == '.') {
                grid[br][bc] = 'O';

                grid[nr][nc] = '@';
                grid[robotRow][robotCol] = '.';

                robotRow = nr;
                robotCol = nc;
            }
        }
    }

    static int gps() {
        int total = 0;

        for (int r = 0; r < grid.length; r++) {
            for (int c = 0; c < grid[r].length; c++) {
                if (grid[r][c] == 'O') {
                    total += 100 * r + c;
                }
            }
        }

        return total;
    }

    public static void main(String[] args) throws Exception {
        readGrid();
        readMoves();
        findRobot();

        for (char move : moves.toCharArray()) {
            move(move);
        }

        System.out.println(gps());
    }
}